### ISSUE TITLE
Ensure Well-Kinded Types

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#999]
A bit of background:

A type system ensures that ill-formed expressions such as:

```
if ("hello world") 123 else true
```

are rejected.

A kind system ensures that ill-formed types such as:

```
List[String, String]
Option[Pure]
Int -> 
{ x: Int, y: Int | Bool }
```

are rejected.

The purpose of this ticket is to ensure that the Flix compiler correctly enforces well-kinded types.

The starting point for this is to change `ast/Type.scala` such that fresh type variables always have the right kind. Specifically, inside Type there is:

`freshTypeVarXXXDeprecated` which should be replaced by calls to either:

- `freshTypeVar` if the type variable represents a proper type (kind Star).
- `freshTypeVarWithKind` passing the appropriate kind (e.g. for records, schemas, etc.)

(Right click-> Find usages in Intellij IDEA should be useful here).

Once this is done, `Unification.scala` lines 270 should enforce well-kindedness:

```scala
    def unifyVar(x: Type.Var, tpe: Type): Result[Substitution, UnificationError] = {
      // The type variable and type are in fact the same.
      if (x == tpe) {
        return Result.Ok(Substitution.empty)
      }

      // The type variable occurs inside the type.
      if (tpe.typeVars contains x) {
        return Result.Err(UnificationError.OccursCheck(x, tpe))
      }

      // HERE
      //if (x.kind != tpe.kind) {
      //  return Result.Err(TypeError.KindError())
      //}

      // We can substitute `x` for `tpe`. Update the textual name of `tpe`.
      if (x.getText.nonEmpty && tpe.isInstanceOf[Type.Var]) {
        tpe.asInstanceOf[Type.Var].setText(x.getText.get)
      }
      Result.Ok(Substitution.singleton(x, tpe))
    }
```

More background material can be found in:

- Types and Programming Languages (chapter 29: Type Operators and Kinding) - note the system here is more expressive than HM.
- Typing Haskell in Haskell (chapter 3 and onwards)
- https://en.wikipedia.org/wiki/Kind_(type_theory)

***I have created a branch named "issue-999-well-kinded" where this work should take place***

**Comment by mlutze** [comment]
Fixed by https://github.com/flix/flix/pull/1265 ?

**Comment by magnus-madsen** [comment]
> 
> 
> Fixed by #1265 ?

Let me think about it. I am not yet convinced that -- during type inference -- an ill-kinded type could arise.
